### ISSUE TITLE
Handling non-UTF-8 strings

### ISSUE THREAD

**Issue by ondrejmirtes** [clue/reactphp-ndjson#21]
Hi, my users that encode their projects in ISO-8859-1 are experiencing this problem: https://github.com/phpstan/phpstan/issues/3956#issuecomment-726613942

![image](https://user-images.githubusercontent.com/104888/99047968-413fff00-2595-11eb-89ce-75586879b925.png)

It's because non-UTF-8 strings can find their way into the JSON encoder.

What's the recommended way of preventing this error? How can I bulletproof-encode non-UTF-8 into UTF-8 before feeding it into reactphp-ndjson?

Thank you!

**Comment by ondrejmirtes** [comment]
Solved: https://github.com/phpstan/phpstan-src/commit/0524fb5c8e060d40ef6de65955e8e4b324d94477

Feel free to close.

**Comment by clue** [comment]
@ondrejmirtes Thanks for reporting this here! I see this is already solved for your use case, but I'm answering here also to help others encountering similar problems.

JSON only supports UTF-8 encoded strings by design, so this is expected behavior by default.

I see you've started using the `JSON_INVALID_UTF8_IGNORE` (or `JSON_INVALID_UTF8_SUBSTITUTE`) constant available as of PHP 7.2.0 to handle invalid UTF-8 data. This sounds like a reasonable approach if you want to ignore invalid bytes in your text input.

As an alternative, you may also want to look into transcoding your input strings to valid UTF-8 encoding. This is relatively straight forward when your input encoding is known (`uft8_encode()`, `iconv()`, `mb_convert_encoding()` come to mind), but can be much harder when you don't control the input encoding.

If you want to transport arbitrary string data without caring about its text encoding (or binary data), native JSON might not be the best choice. You can always encode your input string using base64 encoding before passing it to the JSON parser, but this involves additional overhead (plus effort on the receiving side).

Depending on what you want to achieve, you may also use a different framing format that uses message lengths such as BEncode (https://github.com/clue/reactphp-bencode) or netstrings (https://github.com/clue/reactphp-netstring).

To recap: For your specific use case, it sounds like ignoring invalid bytes might be reasonable. For other use cases, this might not be acceptable. I've worked with a number of projects to evaluate and find the best I/O formats and the answer is really "it depends". If you're interested in professional support, drop me an email and we'll work out the details.

**Comment by ondrejmirtes** [comment]
Thank you very much :)

**Comment by clue** [comment]
For the reference: I've updated my answer to include links to https://github.com/clue/reactphp-bencode and https://github.com/clue/reactphp-netstring as possible alternatives. I hear there will be some updates on both projects not too far in the future… 🤫
